ksqldb-functional-tests/src/test/resources/query-validation-tests/greatest.json

```diff
@@ -0,0 +1,178 @@
+      }
+    },
+    {
+      "name": "test implicit casting to decimal",
+      "statements": [
+        "CREATE STREAM INPUT (ID BIGINT KEY, N1 INT, N2 BIGINT, N3 DECIMAL(50,30)) WITH (kafka_topic='input_topic',value_format='json');",
+
+        "CREATE STREAM OUTPUT AS SELECT ID, GREATEST(N1, N2, N3, null) AS HIGHEST FROM INPUT;"
+
+      ],
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlStatementException",
+        "message": "Function 'greatest' does not accept parameters (INTEGER, BIGINT, DECIMAL(50, 30), null)."
+      }
+    },
+    {
+      "name": "test implicit casting to double",
+      "statements": [
+        "CREATE STREAM INPUT (ID BIGINT KEY, N1 INT, N2 BIGINT, N3 DECIMAL(30,18), N4 DOUBLE) WITH (kafka_topic='input_topic',value_format='json');",
+
+        "CREATE STREAM OUTPUT AS SELECT ID, GREATEST(N1, N2, N3, N4, null) AS HIGHEST FROM INPUT;"
+      ],
+      "inputs": [
+        {"topic": "input_topic", "key": 1, "value": {"N1": -1, "N2": 1000000000000, "N3": 99999.99999}, "N4": 0.0},
+        {"topic": "input_topic", "key": 2, "value": {"N1": -1, "N2": 1, "N3": 55555.555555555555555555}, "N4": 0.0},
+        {"topic": "input_topic", "key": 3, "value": {"N1": 500000, "N2": 1, "N3": 99999.99999}, "N4": 0.0}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "key": 1, "value": {"HIGHEST": 1000000000000}},
+        {"topic": "OUTPUT", "key": 2, "value": {"HIGHEST": 55555.555555555555}},
+        {"topic": "OUTPUT", "key": 3, "value": {"HIGHEST": 500000.0}}
+      ],
+      "post": {
+        "sources": [
+          {"name": "OUTPUT", "type": "stream", "schema": "ID BIGINT KEY, HIGHEST DOUBLE"}
+        ]
+      }
+    },
+    {
+      "name": "test decimal widening",
+      "statements": [
+        "CREATE STREAM INPUT (ID BIGINT KEY, N1 DECIMAL(9,2), N2 DECIMAL (7,6)) WITH (kafka_topic='input_topic',value_format='json');",
+        "CREATE STREAM OUTPUT AS SELECT ID, GREATEST(N1, N2) AS HIGHEST FROM INPUT;"
+      ],
+      "inputs": [
+        {"topic": "input_topic", "key": 1, "value": {"N1": 1234567.89, "N2": 0.123456}, "N4": 0.0}
+      ],
+      "outputs": [
+        {"topic": "OUTPUT", "key": 1, "value": {"HIGHEST": 1234567.89}}
+      ],
+      "post": {
+        "sources": [
+          {"name": "OUTPUT", "type": "stream", "schema": "ID BIGINT KEY, HIGHEST DECIMAL(13,6)"}
+        ]
+      }
+    },
+    {
+      "name": "test greatest with strings",
+      "statements": [
+        "CREATE STREAM INPUT (ID BIGINT KEY, S1 STRING, S2 STRING, S3 STRING, S4 STRING) WITH (kafka_topic='input_topic',value_format='json');",
+        "CREATE STREAM OUTPUT AS SELECT ID, GREATEST(S1, S2, S3, S4, null, null, 'hello') AS HIGHEST FROM INPUT;",
+        "CREATE STREAM OUTPUT2 AS SELECT ID, GREATEST('\"', S1) AS HIGHEST FROM INPUT;",
+        "CREATE STREAM OUTPUT3 AS SELECT ID, GREATEST('\\foo\"', S1) AS HIGHEST FROM INPUT;"
```

Comment:
See above comment regarding the QTT framework not supporting multiple persistent queries. I think just one is fine for testing the function with string inputs.